### ISSUE TITLE
Consider supporting Filter Object

### ISSUE THREAD

**Issue by kanitw** [vega/vega-lite#1376]
@jheer  @domoritz 

Currently our `filter` is using Vega Expression, which is easy for compilation and manual writing, but not easy for static analysis / enumeration.  This will make implementing UI in Polestar / Enumerating in Compass(QL) much easier.  

Should we bring back the support for `filter` in an object form?   I remember we briefly discuss about this in a meeting with @jheer (like in two sentences).  

Now that I think more about CompassQL syntax (and Polestar UI) ... maybe it makes sense to consider it even more closely.  

For example, the following filter
```js
transform: {
  filter: {
    operator: '<',
    operand1:  {field: 'horsepower'}   // alternative names left / x1/ l
    operand2: {value: 9}               // alternative names right / x2 / r
  }
} 
```
might get compiled into `transform: {filter: 'datum.horsepower < 9'}`.  

In some sense, this syntax proposed syntax is more consistent with how we design `FieldDef`s.  

If we want to, we can support 'and' and 'or' like this:  
(This is useful for Polestar, but I don't think we really need it for CompassQL and our new UI prototype.)

```js 
transform: {
  filter: {
    and/or: [
       filter1,  // same syntax as above
       filter2
    ] 
  }
} 
```

Basically here are the TS interfaces.
```js
interface BinaryPredicate {
  operator: '<' | '>' | '==' | ...,
  operand1: FieldDef,
  operand2: FieldDef
}

interface UnaryPredicate {
  operator: ...,   // do we need any unary predicate?
  operand: FieldDef
}

interface And/OrFilter { // there are two separate interfaces
  and/or: Filter[]
}

interface NotFilter {
  not: Filter
}

type Filter = UnaryPredicate | BinaryPredicate | And/OrFilter | NotFilter | string;  // string = expression
```

**Comment by jheer** [comment]
This deserves a more detailed in-person conversation. What you propose is
basically manual specification of an AST. An alternative might be something
more akin to MongoDB's syntax (
https://docs.mongodb.com/manual/tutorial/query-documents/). In any case, we
should discuss.

**Comment by kanitw** [comment]
Sure.  Let's discuss about this on Tuesday.  

Quick Note: the `and` and `or` syntax here is actually borrowed from mongoDB, but the predicate syntax is more consistent with the rest of Vega-Lite.

**Comment by arvind** [comment]
FWIW, in our selections branch, composing selections with and/or follow the MongoDB convention.

**Comment by kanitw** [comment]
@arvind  can you point me to the file that involves `and/or` syntax.  At least I can take a look and make sure it is consistent if we were to do this.  :)

**Comment by arvind** [comment]
Sure, it's in [common.ts](https://github.com/vega/vega-lite/blob/repeat3Selections3/src/compile/common.ts#L73) of good ol' repeat3Selections3.

**Comment by domoritz** [comment]
If it's easy to invoke the vega expression parser to get the AST, I'd say let's stick with string expressions.

**Comment by jheer** [comment]
Along those lines, one option for Compass/GraphScape/etc to consider is
using an *internal* object model that mirrors a vega-expression AST. Then
you can enumerate filters any way you like within that format and run a
codegen pass to generate the string expression for the final output spec.

**Comment by kanitw** [comment]
> Along those lines, one option for Compass/GraphScape/etc to consider is
using an *internal* object model that mirrors a vega-expression AST. Then
you can enumerate filters any way you like within that format and run a
codegen pass to generate the string expression for the final output spec.

I was thinking about this idea as well.  Maybe this is what we should do.

**Comment by kanitw** [comment]
In fact, the original motivation that prompts me to think about this issue is the internal model object.

Still, the common internal object is useful for most projects that use Vega-LIte as internal representation.  
So at least, we should provide it as a helper function for VL users.   

In any case, let's talk more in person :)

**Comment by kanitw** [comment]
From discuss with @jheer, we are talking about something along the line of the following

(Draft from the meeting)

```js 
// each filterPredicate are and-ed 
"filter": [
  {"field": "foo", "values": ["a", "b", "c"], "negate":true},
  {"field": "bar", "range": [-5, 10], "negate":true}
];
// full expression version 
"filter": [
  {"test": "datum.foo == 'Kanit'"}
]
```

Thinking more about it, I think  `equal`, `in` , `min`, `max`

```js 
"filter": [
  {"field": "foo", "equal": "a", "negate": false},   // I want equal separate from in because then we can think about CompassQL's syntax for enumerating the drill down value.  
  {"field": "foo", "in": ["a", "b", "c"], "negate": false},
  {"field": "bar", "min": 5, "max": 10, "negate": false}
];
// full expression version 
"filter": [
  {"test": "datum.foo == 'Kanit'"}
]
```

I will look into how selection works and revise the draft.

**Comment by domoritz** [comment]
I think we should use expressions and give developers a simple method to parse expressions into an AST. For the AST, we can just use the standard https://github.com/estree/estree/blob/master/spec.md.

**Comment by kanitw** [comment]
@domoritz  What's your rationale? 

IMO, adding method just complicates how you can use Vega-lite as a representation for higher-level applications as you need to add parse / unparse phase into your cycle.  Since one of our design goal is to make programmatic generation easy, we should natively support this.  

(The filter as expression is more ideal for human writing.)

**Comment by domoritz** [comment]
> Since one of our design goal is to make programmatic generation easy, we should natively support this.

Yes, that was a design goal but honestly that's not what Vega-Lite has become. I think at its core VL should use expressions for filters/ calculates and so on but we can provide a higher level AST that can compile to expressions. That way we make it easy for people to write/read VL but higher level applications can still reason/ programmatically generate specs.

**Comment by kanitw** [comment]
> Yes, that was a design goal but honestly that's not what Vega-Lite has become.

I disagree with this.  This statement sounds slightly blindsided by iPyVega POV.  
However, this remains a key important design goal of Vega-Lite.  In fact, we still argue for programmatic generation in the Vega-Lite paper. 

That said, expressions make sense for human-written specs.  I think the right solution is to support both of them, which makes sense for each of these two use cases.

**Comment by arvind** [comment]
It seems like while there's general agreement about the _need_ for a JSON-based format for filter predicates, there are differing opinions on where in the stack the JSON -> string conversion should happen. Ultimately, I think predicates will have to be expressed as strings since that is what vega-expression takes as input to construct the necessary JavaScript functions. 

What's less clear to me is why the expression AST should be embedded in the Vega-Lite language? One could imagine adding a compilation utility method to vega-expression that takes an AST, in the format proposed by @kanitw, and returns the corresponding vega-expression string. This would mean that the AST could be used throughout the entire Vega stack. I believe this is/similar to what @jheer proposed earlier, but I'm not sure what the reasons are for discarding it?

**Comment by kanitw** [comment]
> This would mean that the AST could be used throughout the entire Vega stack. 

One thing in the discussion last week is that making `filter` object syntax like AST (like I originally proposed) seems too expressive than typical filter need–hence a simpler syntax for the latest draft.

> What's less clear to me is why the expression AST should be embedded in the Vega-Lite language? One could imagine adding a compilation utility method to vega-expression that takes an AST, in the format proposed by @kanitw, and returns the corresponding vega-expression string.

No matter we introduce non-expression filter as a part of the language or utility function, we need to educate developers who use Vega-Lite as representations how to make filter with the non-expression.
So I don't think it's truly lighter-weight in this sense. 

More importantly, making it a utility function requires programs to add another cycle to sync between internal representation that store the filter as an object (basically an object that looks almost like Vega-LIte except the filter part) and compile that to Vega-Lite every time it changes, which seems like an unnecessary requirement.  If our design goal is to support programatic generation, it is much more natural to just have Vega-Lite as the internal representation.   

Another minor point is that we're extending Vega-Lite with enumeration specifier as in CompassQL.   
Without a property filter syntax that can be easily extended by adding `'?'` (enumeration specifier), this might be an evidence that the current filter syntax doesn't support enumeration that well.

**Comment by arvind** [comment]
> No matter we introduce non-expression filter as a part of the language or utility function, we need to educate developers who use Vega-Lite as representations how to make filter with the non-expression.
So I don't think it's truly lighter-weight in this sense.

Agreed, it would have to be documented regardless. However, my point was that expressing predicates in this way is more complex than a straightforward expression, and is best suited for only one specific use case (enumeration). Making it a utility function, rather than a language feature, would signal that this is an advanced feature provided for convenience and, because it would ultimately produce a string anyways, that expressions remain the "base representation" for filter predicates. 

Another lens through which to evaluate this decision is a separation of concerns. The AST representation makes it easier to enumerate possible filter predicates. Are there other use cases the AST is good for? With enumeration, would that happen within the Vega-Lite compiler or in a high-level tool? My hunch would be that it would be in a higher-level tool, but your comment about extending Vega-Lite with an enumeration specifier gives me pause. I would have expected `?` to feature of CompassQL, be filled in with alternatives at that level, and handed down to Vega-Lite. However, if it is indeed happening in Vega-Lite, then I understand why you are advocating for embedding an AST representation within the Vega-Lite language itself.

**Comment by kanitw** [comment]
>  Are there other use cases the AST is good for? 

This part below, I should made it clear I actually refer to how to implement filter UI in polestar (for example).   

> > making it a utility function requires programs to add another cycle to sync between internal representation that store the filter as an object (basically an object that looks almost like Vega-LIte except the filter part) and compile that to Vega-Lite every time it changes, which seems like an unnecessary requirement. If our design goal is to support programatic generation, it is much more natural to just have Vega-Lite as the internal representation.

---

> I would have expected ? to feature of CompassQL, be filled in with alternatives at that level, and handed down to Vega-Lite. However, if it is indeed happening in Vega-Lite, then I understand why you are advocating for embedding an AST representation within the Vega-Lite language itself.

Sorry for not being clear.  I'm just trying to say that CompassQL's syntax should have very close resemblance to Vega-Lite syntax.  Basically, replacing all ? in CompassQL with actual value should almost much gives Vega-Lite spec.  The fact we cannot do this with `filter` yet kinda left an impression that Vega-lite's filter syntax is inconsistent with the rest of Vega-Lite.   

> Making it a utility function, rather than a language feature, would signal that this is an advanced feature provided for convenience 

I'm not convinced by this argument yet.  It sounds like we're making one of our core design use cases more difficult than it should be.   The fact that it would eventually produce string in Vega doesn't mean we should use string in Vega-Lite.

**Comment by kanitw** [comment]
Here is a more convincing case to think about.  
In analysis, people often want to filter something by month, day of week.  

By making `filter`'s predicate an object.  We can naturally add `timeUnit` to the predicate. 
(The pure Vega expression version does not allow this as it's representing field at a lower-level than Vega-Lite.) 

(If we wanna go to post-aggregate filter #117, same trick can be applied -- but I'm not 100% sure if we should do post-aggregate filter yet.  Looking at Tableau UI, there might be a convincing use-case though.)

**Comment by domoritz** [comment]
> Hmm, but aren't the full gamut of Vega expressions already supported by Vega-Lite's filter? So, couldn't the date/time functions be called directly as part of the expression? Or does Vega-Lite's time unit do more than that?

It does more. For instance, we support `yearmonth`, which extracts the year and the month together with the following expression `datetime(year(datum.date), month(datum.date), 1, 0, 0, 0, 0)`.

**Comment by kanitw** [comment]
While I agree that providing multiple ways of specifying the same thing isn't ideal, I don't think the helper function is the right way to do it.  In fact, if we really have to pick one, I would cut the Vega expression version. 

> When users read the documentation, and browse the examples, they get the sense that the language has a small surface area and is thus is easy to learn. 

This part isn't necessary true.  Either approach requires documentation and thus will be a part of the doc surface area.  In fact, for high-level tool developers, this will drive people away because they will ask why do they have to call these "helper" functions.  

For the part about "they get the sense that the language has small surface area", the Vega expression is actually negative for this case.  Vega Expression is its own language, which has more functionality than VL users need.  (VL users shouldn't have to learn about `signal`, `event`, etc.).
While Vega expression is very close to JS, it's not 100% similar and thus still require learning.  
In context of wrappers in other language like Python / Julia, JS-like syntax isn't ideal for their case.  A natural way for wrappers is to provide specific syntax for filter.  If we want the syntax to be consistent across different wrappers, then it should be a core part of Vega-Lite so different wrapper authors don't make their own different syntax. 

Plus, if we choose to do the helper route, the helper will not work in other language.  For example, Altair might produce JSON object and have to compile the filter part into string.  This helper also has to be written in Python so that the output Python dictionary remains consistent with Vega-Lite's JSON syntax.   

(If you want to call this helper via JS, Altair has to produce a JSON dict that looks almost like VL except the filter part and call this helper every time before calling `vl.compile` -- this seems like a mess.  Simply including filter object in the syntax makes the API call much simpler.)

In hindsight, when we first introduced `filter`, we liked that we use in Vega Expression for `filter` simply because it's easy to implement (from Vega-Lite side), which in some sense means that we were "lazy" back then, not because it's the most natural way to support it.  

I think we should distinguish between the AST filter object and the simple filter object.  
The original AST syntax introduces huge complexity for learning so let's ignore it. 
Meanwhile, the simple filter object is not really harder to learn compared to Vega Expression.  

For example,
`{field: "Horsepower",  "min": 5, "max": 100}` vs `"datum.Horsepower >= 5 && datum.HorsePower <= 100"`

Users might be confused what is `datum.`, which in fact is lower level primitive that comes from Vega.
Meanwhile, the `field` syntax looks more consistent with the rest of Vega-Lite.  

You might argue that `calculate` uses `datum.` too.  That's true.  However, `filter` is something that happens more during rapid exploration (e.g, dynamic filter use case).  `calculate` is more like a change at the schema level (at least thinking about Tableau / iPython model).  If we have to pick to be consistent with core parts of Vega-Lite or just with calculate, I'd definitely pick the core.  

>  As higher-level tools evolve, their compilation process will inevitably grow more complex. Thus, saving a single step may not amount to very much at all.

In Vega & Lyra maybe, so far we haven't really encountered any of these in Polestar and I think this case is not convincing enough for us to force Polestar to start having this problem yet.

**Comment by arvind** [comment]
Thanks, you've sold me!

**Comment by kanitw** [comment]
Thanks for commenting too.  You forced me to really think and write down why this is important :) 

Since this issue is very long discussion about should we do it. I'll close this one and make another issue to discuss the syntax (how should we do it).
